### PR TITLE
Add validation of kubelet and kube-proxy configuration in kubeadm.

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
kubeadm has implemented the support for Kubelet Dynamic Configuration. This PR adds validation on the kubelet configuration.

kube-proxy validation also added in this PR.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref: https://github.com/kubernetes/kubeadm/issues/28

**Special notes for your reviewer**:
/cc @luxas @kubernetes/sig-cluster-lifecycle-pr-reviews

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go [hunk 1]
```diff
@@ -487,3 +489,65 @@ func TestValidateIgnoreChecksErrors(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateKubeletConfiguration(t *testing.T) {
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go]: note: the test case is mainly stolen from the original test case of kubelet validation.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 2]
```diff
@@ -75,6 +78,9 @@ func ValidateMasterConfiguration(c *kubeadm.MasterConfiguration) field.ErrorList
 	allErrs = append(allErrs, ValidateFeatureGates(c.FeatureGates, field.NewPath("feature-gates"))...)
 	allErrs = append(allErrs, ValidateAPIEndpoint(c, field.NewPath("api-endpoint"))...)
 	//allErrs = append(allErrs, ValidateProxy(c, field.NewPath("kube-proxy"))...)
```
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: could you enable the kube-proxy component configuration validation here as well?

File: cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go (hunk 1 above)
Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go]: that's ok I guess :)

**Review by luxas — APPROVED** [review]
1. Should we run the kubeletconfig defaulter in kubeadm defaults.go?
2. Can you enable this for the proxy as well while you're at it?
3. Can we somehow make sure this `kubeletConfig.baseConfig` field doesn't get populated if the feature gate isn't set? It's a pointer so it should work...

**Comment by xiangpengzhao** [comment]
>Should we run the kubeletconfig defaulter in kubeadm defaults.go?

I don't think it's necessary in kubeadm. We just need to care about the fields that we may use in kubeadm, imo.  I add it, anyway.

>Can you enable this for the proxy as well while you're at it?

added.

>Can we somehow make sure this kubeletConfig.baseConfig field doesn't get populated if the feature gate isn't set? It's a pointer so it should work...

the feature gate checking for default added.

@luxas PTAL. Thanks!

**Comment by xiangpengzhao** [comment]
/test pull-kubernetes-unit

`k8s.io/kubernetes/test/integration/garbagecollector TestCRDDeletionCascading 1m21s` :snowflake:

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go [hunk 3]
```diff
@@ -110,7 +112,9 @@ func SetDefaults_MasterConfiguration(obj *MasterConfiguration) {
 	}
 
 	SetDefaultsEtcdSelfHosted(obj)
-	SetDefaults_KubeletConfiguration(obj)
+	if features.Enabled(obj.FeatureGates, features.DynamicKubeletConfig) {
```
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go]: :+1:

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go [hunk 4]
```diff
@@ -200,4 +204,9 @@ func SetDefaults_KubeletConfiguration(obj *MasterConfiguration) {
 	if obj.KubeletConfiguration.BaseConfig.CAdvisorPort == nil {
 		obj.KubeletConfiguration.BaseConfig.CAdvisorPort = utilpointer.Int32Ptr(0)
 	}
+
+	scheme, _, _ := kubeletscheme.NewSchemeAndCodecs()
+	if scheme != nil {
+		scheme.Default(obj.KubeletConfiguration.BaseConfig)
```
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go]: :+1:

**Review by luxas — APPROVED** [review]
/lgtm
Thanks for the quick patch @xiangpengzhao!

**Pull request merged**